### ISSUE TITLE
Analyze the performance of the logs insertion

### ISSUE THREAD

**Issue by pijng** [pijng/moonlogs#10]
## Proposal

Analyze insertion performance as the number of queries/growth of sql base increases.

## Use case

There is a general performance hit on inserting into a standard sqlite database. 

For comparison and further elaboration of this problem I will give below examples of statistics for different time intervals and rpm:

_(Figure 1.1 Insert performance on the first week after release)_
<img width="713" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/a89674ab-53b2-42c6-95fe-bc6c7bb8acdb">

Approximately 1 million records were created during this time.

_(Figure 1.2 Insert performance on the second week)_
<img width="709" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/2c3885b1-d44e-483e-a6af-6f15004b1ca8">

Approximately 3 millions records were created during this time.

We can notice that the average response time has almost doubled with a twofold increase in the number of requests to the service.
Of course, the presence of indexes in the `records` table is of great importance, so the decrease of insertion speed is quite expected as the number of records increases.

The above statistics reflects only a general picture of the situation, as during these periods night time is also taken into account - when requests become much less and, as a consequence, the average response time is distorted to a lesser extent.

_(Figure 2.1 Insert performance during peak hours on the first week)_
<img width="709" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/78dde349-f1a0-48e5-a8ac-56ff8e151f4b">

As far as can be seen, insertion performance during peak hours (3-7pm) on February 26 in the first week of launch averaged 54ms with a 1.1k rpm load.

_(Figure 2.2 Insert performance during peak hours on the second week)_
<img width="711" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/65c72659-6230-4d32-a151-f502c8c0c502">

After a week and a half, insertion performance is already averaging 120ms under a 2.83k rpm load during the same peak hours.

It can be seen that the response time increased by a factor of 2.2 when the load increased by a factor of 2.5.

From this we would like to draw a premature conclusion in the linear increase of response time with the growth of the base and increase of the average rpm.

This may lead to undesirable consequences in the form of slowing down the code from which logs are sent. 
Or an increase in the percentage of data loss due to more frequent timeouts. 

In this connection it is necessary to analyze further performance in detail and confirm/refute this assumption.

If the assumption turns out to be false, it may be possible to get by with general tuning of sqlite database. But it should be taken into account that in the future more and more business processes will be connected to moonlogs, and some of them will be extremely explosive - for example, mass sending of push notifications via FCM. 

But even then, the solution may only be temporary. On the other hand, for such business processes it is quite reasonable to set low `retention_days` for more frequent cleaning of large amount of data.

If the assumption turns out to be correct, you should think about adding a new base adapter. For example, ScyllaDB.

**Comment by pijng** [comment]
Chose mongodb for easier customization on the user side.
It remains to collect metrics over the next few days.

**Comment by pijng** [comment]
<img width="702" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/e63bffc2-9440-4b7a-bb23-2e704af42a0e">

_(Insert performance after changing db adapter from sqlite to mongo)_

Unfortunately, the workloads in the graph do not match the workloads presented earlier, but as a general reference - the average response time significantly decreased immediately after the sqlite → mongodb switch.
The time to read records from the web-interface remained at the same level.
It is worth to observe more.

**Comment by pijng** [comment]
Important info: sqlite actually does a great job of insertion if correctly tuned.

The main problem with sqlite is concurrent writes.
The [sqlite documentation says](https://sqlite.org/wal.html#concurrency):
«Writers merely append new content to the end of the WAL file. Because writers do nothing that would interfere with the actions of readers, writers and readers can run at the same time. However, since there is only one WAL file, **there can only be one writer at a time.**»

That is, even with correctly configured pragmas, we still get bottlenecked because we have a large number of parallel writers.

There is also this precious comment about some fine-grained tuning of sqlite for concurrent writes/reads:
https://github.com/mattn/go-sqlite3/issues/1022#issuecomment-1067353980

Basically it makes sense to try the following:
* make two pools of `*sql.DB`, one with `mode=ro` and one with `mode=rw`;
* set `SetMaxOpenConns(1)` for write pool;
* set `SetMaxOpenConns(runtime.GOMAXPROCS(0) * 2)` for read pool;
* disable `cache=shared`;
* wrap all insert statement to `BEGIN IMMEDITATE ... COMMIT;`.

UPD:
Also it's worth to play a little with [wal_autocheckpoint](https://www.sqlite.org/pragma.html#pragma_wal_autocheckpoint)

The default is 1000 pages, maybe increment it.

**Comment by pijng** [comment]
Implementation of the above points #14

**Comment by pijng** [comment]
<img width="712" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/4c85b985-713d-45a6-83bc-733b46e36ff9">

_(Insert performance)_

The graph is quite chaotic and there are spikes up to ~14-20ms sometimes
And the spikes in response time are not proportional to the spikes on the number of calls.
Presumably it is the go's GC.

Overall not critical.

**Comment by pijng** [comment]
Well, turns out to be pretty good, closing this for now.

<img width="697" alt="image" src="https://github.com/pijng/moonlogs/assets/29716642/495168ef-678d-4d49-9f32-a96a7d218a23">

_(Insert performance with mongo adapter on the 4th day after migration)_
